### DISCUSSION TITLE
How to revalidate data fetched from a database (Next.js 14, app router)

### DISCUSSION THREAD

**Discussion by sebpowell** [vercel/next.js#63662] [Help]
### Summary

Hello, 

I have an eCommerce store for which I am statically generating product / bundle pages – something like this:

```
/app
 /products
  /[slug]
   page.tsx
```

And the code:

```TSX
export const generateStaticParams = async () => {
  const products = await productsService.findAllProducts();

  return products.map((product) => ({
      slug: product.slug,
    })) || [];
};

export default async function Product({
  params,
}: {
  params: { slug: string };
}) {
  const product = await productsService.findBySlug({
    slug: params.slug,
  });

  if (!product) return notFound();

  return <ProductPage product={product} />;
}
```

What I want to be able to do is fetch stock levels on every request – right now, the stock levels are returned via `productsService.findBySlug`, but this requires a rebuild every time the levels change, which is no good. 

According to the docs [here](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers), you can revalidate the cache on every request by adding `const dynamic = 'force-dynamic'`, which I guess would be the right thing to use in this case, as the function is tapping into the database directly – there is no fetch, so I presume the other options outlined in the docs are redundant.

However, this will turn the entire route into a dynamic one, which is not what I want either. Is the right way to handle this to fetch the stock data client-side? Or am I missing a trick?

Any tips would be greatly appreciated, thank you in advance. 

### Additional information

```
Using Next.js 14.1.0 with app router.
```

**Comment by icyJoseph** [comment]
I think people are starting to use this more and more: https://nextjs.org/docs/app/api-reference/functions/unstable_cache

I am not sure if that'll trigger revalidation of a path made during build time though.

Of course, be careful since it is unstable the API might change. Worth noting that it is not the React cache function.

Another option might be https://nextjs.org/docs/app/api-reference/next-config-js/partial-prerendering, and perhaps a combination with the unstable_cache.

**Reply by sebpowell** [reply]
@icyJoseph thank you for the suggestion, I'll have a look. If these don't do the trick, I guess the only option left is to either render the entire page dynamically, or fetch the data client side inside the component itself?

**Comment by Icegreeen** [comment]
You can use both the (Client-Side Fetch) and (Incremental Static Regeneration (ISR)) strategies.

Client-Side Fetch:

 This approach allows you to render the product page statically with the data available at the time of static generation, including current stock levels. It then uses client-side JavaScript to dynamically fetch and update inventory levels as needed. This allows the page to be rendered initially with static data and then dynamically updated with the latest data from the server without having to reload the entire page.

```
export const generateStaticParams = async () => {
  const products = await productsService.findAllProducts();

  return products.map((product) => ({
    params: { slug: product.slug },
  })) || [];
};

export default function Product({
  params,
}: {
  params: { slug: string };
}) {
  const [product, setProduct] = useState(null);
  const [stockLevel, setStockLevel] = useState(null);

  useEffect(() => {
    const fetchProduct = async () => {
      const product = await productsService.findBySlug({
        slug: params.slug,
      });
      setProduct(product);

      // Fetch stock level after product data is fetched
      if (product) {
        const stockLevel = await productsService.fetchStockLevel(product.id);
        setStockLevel(stockLevel);
      }
    };
    fetchProduct();
  }, [params.slug]);
```

Incremental Static Regeneration (ISR):

This strategy uses Next.js Incremental Static Regeneration (ISR) functionality to revalidate only specific data that has changed, such as inventory levels, at defined intervals. This allows you to update the data without rebuilding the entire page, maintaining the benefits of static generation.

```
  export async function getStaticProps({ params }) {
    const product = await productsService.findBySlug({
      slug: params.slug,
    });

    if (!product) {
      return {
        notFound: true,
      };
    }

    return {
      props: {
        product,
      },
      revalidate: 60, // Revalidate every 60 seconds (optional)
    };
  }
```
The first approach, "Client-Side Fetch", handles updating data directly in the client's browser, avoiding the need to revalidate the static page cache. In this case, the user's browser will always make a new request to the server to fetch the latest data, which makes the cache less relevant for updating stock levels.

On the other hand, the second approach, "Incremental Static Regeneration (ISR)", uses Next.js's revalidation feature to update data at defined intervals, ensuring that the static page is periodically updated with the latest data. In this case, the cache can be used to temporarily store data and improve application performance.

@sebpowell

**Reply by sebpowell** [reply]
@Icegreeen thank you! ISR seems like a good solution, but I'm not sure this can be done with the app router? I couldn't find any examples in the docs, but might not have been looking in the right place.

**Reply by sebpowell** [reply]
Ignore me, I found the answer here: https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate – thanks again for the suggestion, going to give it a go.

**Comment by Icegreeen** [comment]
The caching issue does not solve your problem, only if you were having performance problems or have some very specific characteristic, which is not the case.
